File: 백준/Gold/2342. Dance Dance Revolution/Dance Dance Revolution.cc
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const int INF = 1e9;

int cst(int start, int end){
    if(start == 0) return 2;
    if(start == end) return 1;
    if(abs(start-end) == 2) return 4;
    return 3;
}

int main() {
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int dp[5] = {0, INF, INF, INF, INF};
    int pre = 0, cur;
    while(cin >> cur && cur){
        int minv = INF;
        for(int i = 0; i < 5; i++){
            minv = min(minv, dp[i]+cst(i, cur));
            dp[i] += cst(pre, cur);
        }
        dp[pre] = minv;
        pre = cur;
    }
    cout << min({dp[0], dp[1], dp[2], dp[3], dp[4]});
    return 0;
}

File: 백준/Gold/2342. Dance Dance Revolution/Dance Dance Revolution.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

from math import inf

def main():
    cmd = [*map(int, input().split())]
    if not cmd or not cmd[0]: print(0); return
    cmd.pop()
    def cst(start, end):
        if start == 0: return 2
        if start == end: return 1
        if abs(start-end) == 2: return 4
        return 3
    dp = [[inf]*5 for _ in range(5)]
    dp[0][0] = 0
    for c in cmd:
        ndp = [[inf]*5 for _ in range(5)]
        for l in range(5):
            for r in range(5):
                if dp[l][r] == inf: continue
                if c != r: ndp[c][r] = min(ndp[c][r], dp[l][r]+cst(l, c))
                if c != l: ndp[l][c] = min(ndp[l][c], dp[l][r]+cst(r, c))
        dp = ndp
    print(min(map(min, dp)))
    return
if __name__ == "__main__":
    main()